### PR TITLE
- make interval operations scale better

### PR DESCRIPTION
**Pull request by yfarjoun**
### Description

Please explain the changes you made here.
Explain the **motivation** for making this change. What existing problem does the pull request solve?

### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/test/java/htsjdk/samtools/util/IntervalListTest.java [hunk 1]
```diff
@@ -463,16 +469,16 @@ public void testOverlapsSingletonAsListIntervalList(final IntervalList fromLists
                 CollectionUtil.makeCollection(list.iterator()));
     }
 
-    @Test(expectedExceptions = SAMException.class)
+    @Test
     public void testOverlapsEmptyFirstList() {
-        IntervalList.overlaps(Collections.emptyList(), Collections.singletonList(list1));
+        Assert.assertEquals(
+                IntervalList.overlaps(empty, list1), empty);
```
Comment (cmnbroad) [inline src/test/java/htsjdk/samtools/util/IntervalListTest.java]: Why not just add these as additional test cases in the existing data providers ?

File: src/main/java/htsjdk/samtools/util/IntervalList.java [hunk 2]
```diff
@@ -97,9 +106,9 @@ public SAMFileHeader getHeader() {
      * Adds an interval to the list of intervals.
      */
     public void add(final Interval interval) {
-        if (header.getSequence(interval.getContig()) == null) {
-            throw new IllegalArgumentException(String.format("Cannot add interval %s, contig not in header", interval.toString()));
-        }
+        ValidationUtils.validateArg(header.getSequence(interval.getContig()) != null,
+                () -> String.format("Cannot add interval %s, contig not in header", interval.toString()));
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/util/IntervalList.java]: done

File: src/main/java/htsjdk/samtools/util/IntervalList.java [hunk 3]
```diff
@@ -615,46 +616,49 @@ public static IntervalList intersection(final IntervalList list1, final Interval
...
         header.setSortOrder(SAMFileHeader.SortOrder.unsorted);
 
         final IntervalList merged = new IntervalList(header);
+        merged.addall(list1.getIntervals());
 
-        for (final IntervalList in : lists) {
-            SequenceUtil.assertSequenceDictionariesEqual(merged.getHeader().getSequenceDictionary(),
-                    in.getHeader().getSequenceDictionary());
+        SequenceUtil.assertSequenceDictionariesEqual(merged.getHeader().getSequenceDictionary(),
+                list2.getHeader().getSequenceDictionary());
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/util/IntervalList.java]: moved

File: src/main/java/htsjdk/samtools/util/IntervalList.java [hunk 4]
```diff
@@ -615,46 +616,49 @@ public static IntervalList intersection(final IntervalList list1, final Interval
...
+     * Merging does not look for overlapping intervals nor uniquify
+     *
+     * @param lists a list of IntervalList
+     * @return the union of all the IntervalLists in lists.
+     */
+    public static IntervalList concatenate(final Collection<IntervalList> lists) {
+        return lists.stream()
+                .reduce(IntervalList::concatenate)
+                .orElseThrow(() -> new SAMException("Cannot concatenate an empty list of IntervalLists."));
+    }
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/util/IntervalList.java]: good catch. I think I fixed this.

File: src/main/java/htsjdk/samtools/util/IntervalList.java [hunk 5]
```diff
@@ -615,46 +616,49 @@ public static IntervalList intersection(final IntervalList list1, final Interval
...
-            if (intersection == null) {
-                intersection = list;
-            } else {
-                intersection = intersection(intersection, list);
-            }
-        }
-        return intersection;
+        return lists.stream()
+                .reduce(IntervalList::intersection)
+                .orElse(null);
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/util/IntervalList.java]: yes...and I'm not sure how to make it better...

File: src/test/java/htsjdk/samtools/util/IntervalListTest.java (hunk 1 above)
Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/util/IntervalListTest.java]: nice suggestions. done.

File: src/main/java/htsjdk/samtools/util/IntervalList.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/util/IntervalList.java]: I think you made this change in the wrong place (see comment below) instead of here.

**Pull request merged**